File: edu/cmu/hcii/puc/uigen/InterfaceNode.java
/**
 * InterfaceNode.java
 *
 * This is the base class for the set of objects that can be
 * represented in the interface tree.  The interface tree is created
 * from the modified group tree in the dependency graph analysis
 * phase. 
 *
 * @author Jeffrey Nichols
 */

// Package Definition

package edu.cmu.hcii.puc.uigen;


// Import Declarations

import java.awt.Dimension;
import java.awt.Rectangle;

import edu.cmu.hcii.puc.cio.ContainerCIO;


// Class Definition

public abstract class InterfaceNode extends Object {

    //**************************
    // Member Variables
    //**************************

    protected Rectangle m_rBounds;
    protected MultiplePanelNode m_pParent;
    

    //**************************
    // Constructor
    //**************************
    
    public InterfaceNode() {
	m_rBounds = new Rectangle( 0, 0, 0, 0 );
    }


    //**************************
    // Abstract Methods
    //**************************

    public MultiplePanelNode getParent() {

	return m_pParent;
    }

    public Rectangle getBounds() {

	return m_rBounds;
    }

    protected void setParent( MultiplePanelNode pNode ) {
	
	m_pParent = pNode;
    }

    public void setLocation( int nX, int nY ) {

	m_rBounds.x = nX;
	m_rBounds.y = nY;
    }

    public void setSize( int nWidth, int nHeight ) {

	m_rBounds.width = nWidth;
	m_rBounds.height = nHeight;
    }

    public abstract void addComponents( ContainerCIO pContainer );

    public abstract Dimension getPreferredSize( LayoutVariables pVars );

    public abstract void doSizing( LayoutVariables pVars );

    public abstract void doLayout( ContainerCIO pContainer, LayoutVariables pVars );

    /**
     * Inserts the <i>pNode</i> parameter as the parent of this
     * node. <i>pNode</i> becomes the child of the former parent and
     * this node becomes the child of <i>pNode</i>. 
     */
    public void insertAsParent( MultiplePanelNode pNode ) {

	if ( m_pParent != null ) {
	    MultiplePanelNode pParent = m_pParent;
	    pParent.removePanel( this );
	    pParent.addPanel( pNode );
	}

	pNode.addPanel( this );	
    }

    public abstract String toString();
}

File: csharp/UIGeneration/InterfaceNode.cs
using System;
using PUC.CIO;

namespace PUC.UIGeneration
{
	/// <summary>
	/// This is the base class for the set of objects that can be
	/// represented in the interface tree.  The interface tree is created
	/// from the modified group tree in the dependency graph analysis
	/// phase.
	/// </summary>
	public abstract class InterfaceNode
	{
		/*
		 * Member Variables
		 */

		protected ContainerCIO				_container;

		protected System.Drawing.Rectangle	_bounds;
		protected MultiplePanelNode			_parent;

		protected System.Drawing.Size		_minSize;
		protected PUC.Layout.PreferredSize	_prefSize;


		/*
		 * Constructor
		 */

		public InterfaceNode()
		{
			_container = null;

			_bounds = new System.Drawing.Rectangle( 0, 0, 0, 0 );
			_parent = null;

			_minSize = new System.Drawing.Size( 0, 0 );
			_prefSize = null;
		}


		/*
		 * Properties
		 */

		public ContainerCIO Container
		{
			get
			{
				return _container;
			}
			set
			{
				_container = value;
			}
		}


		/*
		 * Member Methods
		 */

		public MultiplePanelNode GetParent()
		{
			return _parent;
		}

		public System.Drawing.Rectangle GetBounds()
		{
			return _bounds;
		}

		public void SetParent( MultiplePanelNode node )
		{
			_parent = node;
		}

		public virtual void SetLocation( int x, int y )
		{
			_bounds.X = x;
			_bounds.Y = y;
		}

		public virtual void SetSize( int width, int height )
		{
			_bounds.Height = height;
			_bounds.Width = width;
		}

		/// <summary>
		/// Inserts the node parameter as the parent of this
		/// node.  node becomes the child of the former parent
		/// and this node becomes this child of node.
		/// </summary>
		/// <param name="node">the node inserted as this node's parent</param>
		public void InsertAsParent( MultiplePanelNode node )
		{
			if ( _parent != null )
			{
				MultiplePanelNode oldparent = _parent;
				oldparent.RemovePanel( this );
				oldparent.AddPanel( node );
			}

			node.AddPanel( this );
		}

		
		/*
		 * Properties
		 */

		public virtual System.Drawing.Size MinimumSize
		{
			get
			{
				return _minSize;
			}
			set
			{
				_minSize = value;
			}
		}

		public virtual PUC.Layout.PreferredSize PreferredSize
		{
			get
			{
				return _prefSize;
			}
			set
			{
				_prefSize = value;
			}
		}


		/*
		 * Abstract Methods
		 */

		public abstract void AddComponents( ContainerCIO container, LayoutVariables vars );

		public abstract void CalculateMinimumSize( LayoutVariables vars );
		public abstract void CalculatePreferredSize( LayoutVariables vars );

		public abstract void DoLayout( LayoutVariables vars );

		public override abstract string ToString();
	}
}
